lib/trie/trie.go

```diff
@@ -550,73 +555,79 @@ func (t *Trie) clearPrefix(curr node, prefix []byte) (n node, wasUpdated bool) {
 
 			curr = handleDeletion(c, c, prefix)
 		}
+		return curr, curr.isDirty()
 	case *leaf:
 		length := lenCommonPrefix(c.key, prefix)
 		if length == len(prefix) {
 			return nil, true
 		}
+		return c, false
 	case nil:
 		return nil, false
 	}
-
-	return curr, curr.isDirty()
+	// This should never happen.
+	return nil, false
 }
 
 // Delete removes any existing value for key from the trie.
 func (t *Trie) Delete(key []byte) error {
 	k := keyToNibbles(key)
-	n, err := t.delete(t.root, k)
-	if err != nil {
-		return err
-	}
-	t.root = n
+	t.root, _ = t.delete(t.root, k)
+	// TODO: Remove error from result.
 	return nil
 }
 
-func (t *Trie) delete(parent node, key []byte) (n node, err error) {
+func (t *Trie) delete(parent node, key []byte) (node, bool) {
 	switch p := parent.(type) {
 	case *branch:
-		length := lenCommonPrefix(p.key, key)
+		// Store the current node and return it, if the trie is not updated.
+		currP := p
+		p = t.maybeUpdateBranchGeneration(p)
```

Comment:
`currP` and `p` are both pointers to the same node, so if you update data in `p` then `currP` gets updated also, does this result in the path's generation being updated even if nothing is deleted?

Reply:
`currP` and `p` will be different if the generation is updated. So as we travel down the path we create a copy of every node but then if nothing is deleted we should discard the copy and return the original node. Hence, I am storing the original node in a temporary variable `currP`. 

If `p` and `currP`  both belong to the same generation then they will point to the same node and it will follow the old code logic. In this case the nodes won't be updated if there is nothing to delete.